src/lib/comms/sol-netctl-impl-connman.c

```diff
@@ -0,0 +1,1133 @@
+        _free_connman_service(service);
+
+    sol_ptr_vector_clear(&_ctx.service_vector);
+    sol_monitors_clear(&_ctx.service_ms);
+    sol_monitors_clear(&_ctx.manager_ms);
+    sol_monitors_clear(&_ctx.error_ms);
+
+    _ctx.connman_state = SOL_CONNMAN_STATE_UNKNOWN;
+}
+
+static int
+_match_properties_changed(sd_bus_message *m, void *userdata,
+    sd_bus_error *ret_error)
+{
+    const char *interface;
+    int r;
+
+    SOL_NULL_CHECK(_ctx.properties_changed, -EINVAL);
+
+    interface = sd_bus_message_get_interface(m);
+    SOL_NULL_CHECK(interface, -EINVAL);
+
+    if (!strstartswith(interface, "net.connman."))
+        return -EINVAL;
+
+    if (sd_bus_message_is_signal(m, "net.connman.Manager",
+        "ServicesChanged")) {
+        if (_ctx.service_slot)
+            return -EINVAL;
+
+        services_list_changed(m);
+    } else if (sd_bus_message_is_signal(m, "net.connman.Manager",
+        "PropertyChanged")) {
+        char *str = NULL;
+
+        if (_ctx.manager_slot)
+            return -EINVAL;
+
+        r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &str);
+        SOL_INT_CHECK(r, < 0, r);
+        if (streq(str, "State")) {
+            r = get_manager_properties(m);
+            SOL_INT_CHECK(r, < 0, r);
+            call_manager_monitor_callback();
+        } else {
+            SOL_DBG("Ignored changed property: %s", str);
+            r = sd_bus_message_skip(m, "v");
+            SOL_INT_CHECK(r, < 0, r);
+        }
+    }
+
+    return 0;
+}
+
+static int
+dbus_service_add_monitor(
+    sol_connman_service_monitor_cb cb, const void *data)
+{
+    int r;
+    struct sol_monitors_entry *e;
+    char matchstr[256];
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+
+    r = sol_monitors_find(&_ctx.service_ms, (sol_monitors_cb_t)cb, data);
+    if (r < 0) {
+        e = sol_monitors_append(&_ctx.service_ms, (sol_monitors_cb_t)cb, data);
+        SOL_NULL_CHECK(e, -ENOMEM);
+    }
+
+    if (_ctx.properties_changed)
+        return 0;
+
+    r = snprintf(matchstr, sizeof(matchstr),
+        "type='signal',"
+        "interface='net.connman.Manager'");
+    SOL_INT_CHECK(r, < 0, r);
+
+    r = sd_bus_add_match(bus, &_ctx.properties_changed, matchstr,
+        _match_properties_changed, NULL);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+static int
+dbus_service_del_monitor(
+    sol_connman_service_monitor_cb cb, const void *data)
+{
+    int r;
+
+    r = sol_monitors_find(&_ctx.service_ms, (sol_monitors_cb_t)cb, data);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return sol_monitors_del(&_ctx.service_ms, r);
+}
+
+SOL_API int
+sol_connman_add_service_monitor(
+    sol_connman_service_monitor_cb cb, const void *data)
+{
+    int r;
+
+    r = sol_connman_init_lazy();
+    SOL_INT_CHECK_GOTO(r, < 0, fail_init);
+
+    r = dbus_service_add_monitor(cb, data);
+    SOL_INT_CHECK_GOTO(r, < 0, fail);
+
+    r = dbus_connection_get_service_properties();
```

Comment:
only need to do this if this is the first, otherwise it was already done and the properties are being monitored. You can return this from `dbus_service_add_monitor()` (ie: return 1 in the first time, 0 on the others)